Implement CLI Parameter to Print Specified Object Types from OpenAPI Spec

wldeh opened the pull request:
Closes #6 . Addresses the qualifying task for GSoC'23 by implementing the functionality to parse an OpenAPI spec from a given URL and print the list of objects identified by the type specified as a CLI parameter. The supported object types include schemas, responses, parameters, examples, requestBodies, headers, securitySchemes, links, callbacks, and pathItems.

wldeh committed:
initial commit

bavulapati commented:
@wldeh Please take a look at the [comment](https://github.com/postman-open-technologies/openapi-componentizer/issues/6#issuecomment-1488251123) on the issue #7 

Also, you can link issues and PRs by using references in the PR description. https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

bavulapati commented:
Also, take a look at https://www.conventionalcommits.org/en/v1.0.0/

wldeh committed:
refactor: Move OpenAPI Parser code to a separate directory

wldeh commented:
Thank you for your detailed feedback, @bavulapati. I reviewed the comment on issue #7 and am making the necessary changes. My condolences for not mentioning the issue in my pr. I have always strived to follow the guidelines outlined by Conventional Commits when making commits to my projects and appreciate the link you provided and will make sure to reference it in the future. Thanks again for your feedback!

wldeh committed:
refactor!: major rewrite with updated functionality

This commit includes a major rewrite of the project based on updated understanding of the qualifying task.

wldeh committed:
fix(cli): Add error messages for missing and invalid options in CLI arguments

This commit adds error messages to the CLI options validation function so that users will be notified if they're missing a required option or providing an invalid value.

wldeh committed:
feat: Validate and handle errors for OpenAPI spec fetcher

The commit adds a new 'isValidUrl' function to check if the given url is valid.
It also validates the URL before fetching the OpenAPI spec, throwing an error if the URL is invalid. If the response is not ok, another error is thrown. The commit message reflects the addition of valid URL validation and handling of errors during the OpenAPI spec fetch operation.

wldeh committed:
docs: add TypeDoc comments and improve code documentation

- Add TypeDoc comments to all functions and types
- Include additional comments to clarify logic and usage

wldeh committed:
docs: add README and generate TypeDoc documentation

- Create a README with usage instructions and examples
- Generate TypeDoc documentation (better code understanding and maintenance)

wldeh committed:
feat(testing): add live test, install Jest, and configure Jest

- Add live test
- Install Jest as a testing framework
- Add Jest configuration to package.json
- Create jest.config.js for Jest settings

wldeh committed:
feat(parser): Revamp parser.ts for improved type handling and update docs

Completely revamp the OpenAPI parser in parser.ts to handle types more accurately and efficiently. Update documentation to reflect changes.

bavulapati commented on openapi-parser/.husky/pre-commit:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
```
Is this auto generated by husky?

wldeh replied:
Yes, this is autogenerated by husky via the `npx husky add .husky/pre-commit "npm run pre-commit"` command

bavulapati commented on openapi-parser/docs/README.md:
Is this auto-generated?
Looks like main README.

wldeh replied:
Yes @bavulapati , the `openapi-parser/docs/README.md` file is the original README autogenerated and modified by Typedoc to include a reference to [Exports](modules.md) at the top.

bavulapati commented on openapi-parser/docs/README.md:
```diff
@@ -0,0 +1,52 @@
+
+## Usage
+
+Use `ts-node` to run the CLI tool:
```
Nice to have:  tsc can be put to work and run the js using node.

wldeh replied:
Hello @bavulapati. I have made a recent commit [38f0ef5](https://github.com/postman-open-technologies/openapi-componentizer/pull/7/commits/38f0ef520fba67b8ee2f7ed5e33a488a471d948f) that addresses this. Please do not hesitate to make any other suggestions.

bavulapati commented on openapi-parser/README.md:
````diff
@@ -0,0 +1,50 @@
+```bash
+git clone https://github.com/postman-open-technologies/openapi-componentizer.git
+cd openapi-parser
+npm install
````
`npm i` fails with
```
❯ npm i

> openapi-componentizer@0.0.1 prepare
> husky install

.git can't be found (see https://git.io/Jc3F9)
npm ERR! code 1
npm ERR! path /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser
npm ERR! command failed
npm ERR! command sh -c husky install

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/balakrishna/.npm/_logs/2023-03-30T13_36_34_943Z-debug-0.log
```

wldeh replied:
I have addressed the issue you mentioned in my recent commit (https://github.com/postman-open-technologies/openapi-componentizer/pull/7/commits/072bd7895a6aefbee63e4111fc9ddb86e7423397). Feel free to let me know if you encounter any further issues after pulling and thank you for bringing this to my attention.

bavulapati reviewed (changes requested):
Requesting changes as I couldn't test it due to npm install failure.

As you are using handful of utilities, it might be worth describing the way you use them and a way to navigate the code accordingly.

CMIIW, I notice you include tests. It's worth providing a command to run them.

wldeh commented:
> Requesting changes as I couldn't test it due to npm install failure.
> 
> As you are using handful of utilities, it might be worth describing the way you use them and a way to navigate the code accordingly.
> 
> CMIIW, I notice you include tests. It's worth providing a command to run them.

Hello @bavulapati! Yes, I can definitely make the documentation as comprehensive as possible. I’ll push a commit tonight that explains each dependency and why its frequently used in a TypeScript project like this. Additionally, I’ll make sure to include instructions in the README on how to run tests and why I chose to use Jest. You can run the existing tests right now by executing the test script defined in the package.json file. I’m still working on writing robust unit tests, which takes some time. Your feedback is much appreciated.

wldeh committed:
fix: update prepare script to correctly set up Husky

The prepare script in package.json has been updated to fix the issue with Husky not finding the .git directory. The new script changes the directory to the appropriate location before running husky install. This should resolve the error encountered during
pm i.

wldeh committed:
feat: update build process to use tsup and improve scripts

Based on the suggestion to use 'tsc' for building and running the project with 'node', the build process has been updated to use 'tsup' for better performance. The 'package.json' scripts have been refactored including the addition of a 'clean' script and using 'npm-run-all' for running scripts in sequence. This improves the overall build process, developer experience, and maintainability.

bavulapati commented:
I think `npm run build` need to be run before running the cli, at the moment.
Ran into the following 

```
ESM Build start
✘ [ERROR] Could not resolve "node-fetch"

    src/openapi/fetcher.ts:1:18:
      1 │ import fetch from 'node-fetch'
        ╵                   ~~~~~~~~~~~~

  You can mark the path "node-fetch" as external to exclude it from the bundle, which will remove
  this error.

ESM Build failed
Error: Build failed with 1 error:
src/openapi/fetcher.ts:1:18: ERROR: Could not resolve "node-fetch"
    at failureErrorWithLog (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1636:15)
    at /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1048:25
    at runOnEndCallbacks (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1471:45)
    at buildResponseToResult (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1046:7)
    at /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1075:16
    at responseCallbacks.<computed> (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:697:9)
    at handleIncomingPacket (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:752:9)
    at Socket.readFromStdout (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:673:7)
    at Socket.emit (node:events:512:28)
    at addChunk (node:internal/streams/readable:324:12)
```

wldeh commented:
> I think `npm run build` need to be run before running the cli, at the moment. Ran into the following
> 
> ```
> ESM Build start
> ✘ [ERROR] Could not resolve "node-fetch"
> 
>     src/openapi/fetcher.ts:1:18:
>       1 │ import fetch from 'node-fetch'
>         ╵                   ~~~~~~~~~~~~
> 
>   You can mark the path "node-fetch" as external to exclude it from the bundle, which will remove
>   this error.
> 
> ESM Build failed
> Error: Build failed with 1 error:
> src/openapi/fetcher.ts:1:18: ERROR: Could not resolve "node-fetch"
>     at failureErrorWithLog (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1636:15)
>     at /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1048:25
>     at runOnEndCallbacks (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1471:45)
>     at buildResponseToResult (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1046:7)
>     at /Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:1075:16
>     at responseCallbacks.<computed> (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:697:9)
>     at handleIncomingPacket (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:752:9)
>     at Socket.readFromStdout (/Users/balakrishna/projects/postman-open-technologies/openapi-componentizer/wldeh/openapi-componentizer/openapi-parser/node_modules/esbuild/lib/main.js:673:7)
>     at Socket.emit (node:events:512:28)
>     at addChunk (node:internal/streams/readable:324:12)
> ```

Hello @bavulapati. I was swamped with work today and couldn’t provide an update. I’ll follow up with you by the 31st and look to re-request. Thank you for your timely feedback; I appreciate it greatly.

wldeh committed:
fix: update build and start scripts, add node-fetch

wldeh committed:
fix: run pre-commit script in openapi-parser directory

wldeh committed:
docs: Enhance README with detailed dependency explanations and code navigation

wldeh committed:
docs: Add TODO section to README with planned improvements

wldeh commented:
Hello @bavulapati! I wanted to inform you that I have just addressed most of the changes that you mentioned. Please let me know if you have any questions or concerns. Thank you!
